kernel/timeout.c

```diff
@@ -169,29 +162,46 @@ void z_clock_announce(s32_t ticks)
 #endif
 
 	k_spinlock_key_t key = k_spin_lock(&timeout_lock);
+	struct _timeout *t = first();
 
-	announce_remaining = ticks;
+	curr_tick += ticks;
 
-	while (first() != NULL && first()->dticks <= announce_remaining) {
-		struct _timeout *t = first();
-		int dt = t->dticks;
+	if (t) {
+		sys_dlist_t ready;
+		sys_dnode_t *node;
+		s32_t announce_remaining = ticks;
 
-		curr_tick += dt;
-		announce_remaining -= dt;
-		t->dticks = 0;
-		remove_timeout(t);
+		while (t) {
+			if (t->dticks > announce_remaining) {
+				break;
+			}
+			announce_remaining -= t->dticks;
+			t = next(t);
+		}
 
-		k_spin_unlock(&timeout_lock, key);
-		t->fn(t);
-		key = k_spin_lock(&timeout_lock);
-	}
+		sys_dlist_init(&ready);
+		if (t == NULL) {
+			sys_dlist_catenate(&ready, &timeout_list);
+		} else {
+			sys_dlist_split(&timeout_list, &t->node, &ready);
+			t->dticks -= announce_remaining;
+		}
 
-	if (first() != NULL) {
-		first()->dticks -= announce_remaining;
-	}
+		node = sys_dlist_peek_head(&ready);
+		if (node) {
+			k_spin_unlock(&timeout_lock, key);
+			while (node) {
```

Comment:
`do { } while (node = sys_dlist_peek_head(&ready))`
In my opinion such code will make concept easier to understand.